src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,172 @@
+        end
+        api_client = VSphereAutomation::ApiClient.new(configuration)
+        api_client.default_headers['Authorization'] = configuration.basic_auth_token
+        session_api = VSphereAutomation::CIS::SessionApi.new(api_client)
+        session_id = session_api.create('').value
+        api_client.default_headers['vmware-api-session-id'] = session_id
+        api_client
+      end
+
+      def initialize(api_client)
+        @api_client = api_client
+      end
+
+      private def tag_assoc_api
+        @tag_assoc_api ||= VSphereAutomation::CIS::TaggingTagAssociationApi.new(@api_client)
+      end
+
+      private def tagging_tag_api
+        @tagging_tag_api ||= VSphereAutomation::CIS::TaggingTagApi.new(@api_client)
+      end
+
+      private def tagging_category_api
+        @tagging_category_api ||= VSphereAutomation::CIS::TaggingCategoryApi.new(@api_client)
+      end
+
+      def retrieve_category_id(category_name, category_ids)
+        category_ids.each do |category_id|
+          result = tagging_category_api.get(category_id)
+          return result.value.id if category_name == result.value.name
+        end
+        nil
+      end
+
+      def retrieve_tag_id(tag_name, tag_id_list)
+        tag_id_list.each do |tag|
+          result = tagging_tag_api.get(tag)
+          return result.value.id if tag_name == result.value.name
+        end
+        nil
+      end
+
+      def vm_association?(target_category_id)
+        category_information = tagging_category_api.get(target_category_id)
+        category_assoc_types = category_information.value.associable_types
+        return true if category_assoc_types.empty? || category_assoc_types.include?("VirtualMachine")
+        false
+      end
+
+      def single_category_cardinality?(target_category_id)
+        category_information = tagging_category_api.get(target_category_id)
+        return true if category_information.value.cardinality == "SINGLE"
+        false
+      end
+
+      def create_tag_hash(vm_config_tags)
+        tag_hash = Hash.new
+        vm_config_tags.each do |vm_config_tag|
+
+          if vm_config_tag["category"].nil?
+            logger.warn( "Missing category content, skip processing this category-tag pair." )
+            next
+          end
+          if vm_config_tag["category"].empty?
+            logger.warn( "Empty category found in cloud config ,skip processing this category-tag pair." )
+            next
+          end
+          if vm_config_tag["tag"].nil?
+            logger.warn( "Missing tag in category '#{vm_config_tag["category"]}', skip attaching this tag." )
+            next
+          end
+          if vm_config_tag["tag"].empty?
+            logger.warn( "Empty tag found in category '#{vm_config_tag["category"]}', skip attaching this tag." )
+            next
+          end
+          tag_hash[vm_config_tag["category"]] ||= []
+          tag_hash[vm_config_tag["category"]] << vm_config_tag["tag"]
```

Comment:
If `vm_config_tag["tag"]` is an array, you will end up with a nested array. Either correct this behavior using `flatten` or state clearly in `docs-bosh` PR that tags cannot be an array.

Reply:
Hi @EleanorRigby , I will state clearly in `docs-bosh` in this version. Might try to support tag input as arrays in the PR if we decide to make vm create failed when get tagging failures.